codePost_api/helpers.py

```diff
@@ -184,7 +189,45 @@ class UploadModes(_DocEnum):
 # Core upload logic
 ###########################################################################################
 class UploadError(RuntimeError):
-    pass
+
+    def __init__(self, message, api_key=None, assignment_id=None, submission_id=None, file_ids=None, *args, **kwargs):
+        super(UploadError, self).__init__(message, *args, **kwargs)
+        
+        self.api_key = api_key
+        self._assignment_id = assignment_id
+        
+        # Initialize object lists
+        self._submission_ids = [] if submission_id == None else [ submission_id ]
+        self._file_ids = [] if file_ids == None else file_ids
+
+        # Trace of operations undertaken by the cleanup method
+        self._trace = ""
+    
+    def force_cleanup(self):
+
+        self._trace += "<<"
+        
+        for f_id in self._file_ids:
+            try:
+                delete_file(api_key=self.api_key, file_id=f_id)
+            except:
+                self._trace += "!file({}); ".format(f_id)
+            else:
+                self._trace += "-file({}); ".format(f_id)
+        
+        for s_id in self._submission_ids:
+            try:
+                delete_submission(api_key=self.api_key, submission_id=s_id)
+            except:
+                self._trace += "!submission({}); ".format(s_id)
+            else:
+                self._trace += "-submission({}); ".format(s_id)
+        
```

Comment:
This isn't well documented, but deleting a submission cascades the delete operation to all child files. So, to delete a submission and all of its files, you just need to delete the submission. Relying on the API-side delete cascade here would simplify `force_cleanup`. Note that there can be no issue with "dangling files" (i.e. files without a parent submission), because our database constraints do not allow such files (the `submission` field of the `File` model is not optional).

Reply:
That's what I suspected. I think this is great to insure consistency of the database at all times, and to simplify operations, for instance, through the Django admin panel.

But I see at least one advantage in proceeding this way (and I wonder what you think of this rationale): It is useful to do this to be able to explain to the user the actions that have been taken. (Analogue of the Django confirmation dialog that explains that when you remove a model, it removes all the associated models.)

@jamesaevans Thoughts?

Reply:
I agree with you here, too. Deleting the comments first and then the category is a more self-documenting operation, which I think is valuable even if it comes at the cost of extra API requests. The key here to me is that this code / library is meant to be viewed, not just treated as implementation.

Reply:
Just have to be careful to delete the comments first, since if the category gets deleted first, the requests to delete comments will return 404.

Reply:
@jamesaevans Just to clarify, are you saying that the `UploadError` class should also handle comments? Currently it only handles files and submissions, because those are the only objects that this script operates on itself. But I guess it is not good practice given that the submission could have comments as well.

I am not sure how comment categories are relevant here, since IMHO, they are an assignment property, not a submission property?

Reply:
Sorry, I was working on some code relating to the rubric items, which present the same inheritance question as Files and Submissions! Substitute `submission` for `category` and `file` for `comment` :)

```
I agree with you here, too. Deleting the **files** first and then the **submissions** is a more self-documenting operation, which I think is valuable even if it comes at the cost of extra API requests. The key here to me is that this code / library is meant to be viewed, not just treated as implementation.
```

Reply:
Understood. I added a comment to the exception's cleanup code to document this nuance.